File: First Missing Positive.cpp
class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int i = 0;
        while(i < n) {
            if (A[i] > 0 && A[i] <= n) {
                if (A[i] != i+1 && A[A[i]-1] != A[i]) {
                    int tmp = A[A[i]-1];
                    A[A[i]-1] = A[i];
                    A[i] = tmp;
                } else {
                    i++;
                }
            } else {
                i++;
            }
        }
        
        for (int i = 0; i < n; i++) {
            if (A[i] != i+1) return i+1;
        }
        
        return n+1;
    }
};

File: First Missing Positive.cs
public class Solution {
    public int FirstMissingPositive(int[] nums) {
        if (nums == null || nums.Length == 0) {
            return 1;
        }
        
        for (var i = 0; i < nums.Length; i++) {
            while (nums[i] != i + 1) {
                if (nums[i] <= 0 || nums[i] >= nums.Length || nums[i] == nums[nums[i] - 1]) {
                    break;
                }
                
                var temp = nums[i];
                nums[i] = nums[temp - 1];
                nums[temp - 1] = temp;
            }
        }
        
        for (var i = 0; i < nums.Length; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        
        return nums.Length + 1;
    }
}
